ironfish/src/wallet/wallet.ts

```diff
@@ -484,9 +483,47 @@ export class Wallet {
       }
     })
 
-    for (const account of accounts) {
-      const shouldDecrypt = await this.shouldDecryptForAccount(blockHeader, account)
+    const shouldDecryptAccounts = (
+      await Promise.all(
+        accounts.map((a) =>
+          this.shouldDecryptForAccount(blockHeader, a).then(
+            (shouldDecrypt): [Account, boolean] => [a, shouldDecrypt],
+          ),
+        ),
+      )
+    )
+      .filter((v) => v[1])
+      .map((v) => v[0])
+
+    const shouldDecryptAccountIds = new Set(shouldDecryptAccounts.map((a) => a.id))
+
+    const decryptedTransactions = await Promise.all(
+      transactions.map(({ transaction, initialNoteIndex }) =>
+        this.decryptNotes(transaction, initialNoteIndex, false, shouldDecryptAccounts).then(
+          (r) => ({
+            result: r,
+            transaction,
+          }),
+        ),
+      ),
+    )
 
+    // account id -> transaction hash -> Array<DecryptedNote>
+    const decryptedNotesMap: Map<string, Map<string, Array<DecryptedNote>>> = new Map()
+    for (const { transaction, result } of decryptedTransactions) {
+      for (const [accountId, decryptedNotes] of result) {
+        if (!decryptedNotes.length) {
+          continue
+        }
+
+        const accountTxnsMap =
+          decryptedNotesMap.get(accountId) ?? new Map<string, Array<DecryptedNote>>()
+        accountTxnsMap.set(transaction.hash().toString('base64'), decryptedNotes)
```

Comment:
why base64 instead of hex?

Reply:
not sure if it would be more or less efficient, but we could also use a `BufferMap` to avoid converting the hashes to strings

Reply:
I'll switch to using BufferMap here, it's more clear than making sure the encoding is same on both ends. FWIW under the hood BufferMap just converts Buffer keys using `Buffer.toString('binary')`